### PR TITLE
feat: Add calendar components form validation

### PR DESCRIPTION
**Pull request by JKMarkowski**
#### Please provide a link to the associated issue.
fixes: https://github.com/SAP/fundamental-ngx/issues/998
#### Please provide a brief summary of this pull request.
I added Validator support for calendar/date/datetime components. It forced me to change handling the invalid data. Before on invalid data, model wasn't changed and the data, until it's valid. That approach caused, that the wrong data wasn't shown on examples. 
Right now invalid data is stored on component model and is reflected on forms/models on components that are using calendar components. Tests has been changed also.
Additionally there are provided some basic UX/Visual fixes.
#### If this is a new feature, have you updated the documentation?
I added examples of usage on all of calendar related components.

### CODE REVIEW

File: library/src/lib/calendar/calendar.component.ts [hunk 1]
```diff
@@ -372,6 +391,35 @@ export class CalendarComponent implements OnInit, ControlValueAccessor {
...
+                (
+                    this.selectedRangeDate.start &&
+                    this.selectedRangeDate.start instanceof FdDate &&
+                    this.selectedRangeDate.start.isDateValid()
+                ) && (
+                    this.selectedRangeDate.end &&
+                    this.selectedRangeDate.end instanceof FdDate &&
+                    this.selectedRangeDate.start.isDateValid()
+                )
+                ;
```
Comment (MattL75) [inline library/src/lib/calendar/calendar.component.ts]: probably a bit overkill to put this on its own line

File: library/src/lib/calendar/calendar.component.ts [hunk 2]
```diff
@@ -225,15 +230,29 @@ export class CalendarComponent implements OnInit, ControlValueAccessor {
...
                 }
             }
         }
         this.isValidDateChange.emit(valid);
     }
 
+    /**
+     * @hidden
+     * Function that implements Validator Interface, adds validation support for forms
+     * */
```
Comment (MattL75) [inline library/src/lib/calendar/calendar.component.ts]: documentation weirdly formatted, the last like can have only */

File: library/src/lib/date-picker/date-picker.component.ts [hunk 3]
```diff
@@ -273,23 +293,41 @@ export class DatePickerComponent implements ControlValueAccessor {
...
         if (!selected) {
             this.inputFieldDate = '';
             return;
         }
-        if (this.type.toLocaleLowerCase() === 'single') {
+        if (this.type === 'single') {
+            /**
+             * For single mode, if the date is invalid, model is changed, but it does not refresh currently
+             * displayed day view, or input field text
+             * */
```
Comment (MattL75) [inline library/src/lib/date-picker/date-picker.component.ts]: */

File: library/src/lib/date-picker/date-picker.component.ts [hunk 4]
```diff
@@ -273,23 +293,41 @@ export class DatePickerComponent implements ControlValueAccessor {
...
+                this.calendarComponent.setCurrentlyDisplayed(this.selectedDate);
+                this.inputFieldDate = this.dateAdapter.format(selected);
+            }
+
         } else {
+
+            /**
+             * For range mode, if the date is invalid, model is changed, but it does not refresh currently
+             * displayed day view, or input field text
+             * */
```
Comment (MattL75) [inline library/src/lib/date-picker/date-picker.component.ts]: */

File: library/src/lib/date-picker/date-picker.component.ts [hunk 5]
```diff
@@ -273,23 +293,41 @@ export class DatePickerComponent implements ControlValueAccessor {
...
-                this.calendarComponent.setCurrentlyDisplayed(this.selectedRangeDate.start);
-                this.inputFieldDate = this.dateAdapter.format(selected.start) +
-                    this.dateAdapter.rangeDelimiter + this.dateAdapter.format(selected.end)
-                ;
+
+                if (this.isModelValid()) {
+                    this.calendarComponent.setCurrentlyDisplayed(this.selectedRangeDate.start);
+                    this.inputFieldDate = this.dateAdapter.format(selected.start) +
+                        this.dateAdapter.rangeDelimiter + this.dateAdapter.format(selected.end)
+                    ;
```
Comment (MattL75) [inline library/src/lib/date-picker/date-picker.component.ts]: better to put on previous line

File: library/src/lib/date-picker/date-picker.component.ts [hunk 6]
```diff
@@ -302,47 +340,82 @@ export class DatePickerComponent implements ControlValueAccessor {
...
-            if (this.type === 'single') {
-                const fdDate = this.dateAdapter.parse(date);
+        /** Case when there is single mode */
+        if (this.type === 'single') {
+
+            const fdDate = this.dateAdapter.parse(date);
+
+            /** Check if dates are equal, if dates are the same there is no need to make any changes
+             * Date in model is changed no matter if the parsed date fro string is valid or not.
+             * */
```
Comment (MattL75) [inline library/src/lib/date-picker/date-picker.component.ts]: */

File: library/src/lib/date-picker/date-picker.component.ts [hunk 7]
```diff
@@ -302,47 +340,82 @@ export class DatePickerComponent implements ControlValueAccessor {
...
+            if (!CalendarService.datesEqual(fdDate, this.selectedDate)) {
                 this.isInvalidDateInput = !fdDate.isDateValid();
+                this.selectedDate = fdDate;
+                this.onChange(this.selectedDate);
+                this.selectedDateChange.emit(this.selectedDate);
 
-                // If is correct and data is not exactly the same
-                if (!this.isInvalidDateInput && !CalendarService.datesEqual(fdDate, this.selectedDate)) {
-                    this.selectedDate = fdDate;
+                /** Check if date is valid, if it's not, there is no need o refresh calendar */
```
Comment (MattL75) [inline library/src/lib/date-picker/date-picker.component.ts]: no need to

File: library/src/lib/date-picker/date-picker.component.ts [hunk 8]
```diff
@@ -302,47 +340,82 @@ export class DatePickerComponent implements ControlValueAccessor {
...
 
+            /** Case when there is range mode */
+        } else {
+            const currentDates = date.split(this.dateAdapter.rangeDelimiter);
+            const firstDate = this.dateAdapter.parse(currentDates[0]);
+            const secondDate = this.dateAdapter.parse(currentDates[1]);
+
+            /** Check if dates are equal, if dates are the same there is no need to make any changes
+             * Date in model is changed no matter if the parsed dates from string are valid or not.
+             * */
```
Comment (MattL75) [inline library/src/lib/date-picker/date-picker.component.ts]: Start description after first line, also */ last line

File: library/src/lib/datetime-picker/datetime-picker.component.ts [hunk 9]
```diff
@@ -311,23 +330,27 @@ export class DatetimePickerComponent implements OnInit, OnDestroy, ControlValueA
...
             this.setInput(this.date);
         }
     }
 
     /**
      * @hidden
-     * Method that is triggered by events from calendar component, when there is selected date changed
+     * Method that is triggered by events from calendar component, when there is selected date changed.
+     * If invalid time model is detected, it takes time model data from TimeComponent.
      * */
```
Comment (MattL75) [inline library/src/lib/datetime-picker/datetime-picker.component.ts]: */

File: library/src/lib/datetime-picker/datetime-picker.component.ts [hunk 10]
```diff
@@ -354,31 +377,38 @@ export class DatetimePickerComponent implements OnInit, OnDestroy, ControlValueA
     /**
      * @hidden
      * Method, which is responsible for transforming string to datetime, depending on type or
-     * validation the results are different. It also changes to state of isInvalidDateInput
+     * validation the results are different. It also changes to state of isInvalidDateInput.
      * */
```
Comment (MattL75) [inline library/src/lib/datetime-picker/datetime-picker.component.ts]: */

**Comment by JKMarkowski** [comment]
Comments solved.

**Comment by InnaAtanasova** [comment]
The `compact` date picker does not show the invalid state

Screen Shot 2019-08-08 at 7 33 45 PM

**Comment by InnaAtanasova** [comment]
The first selected date in the range calendar has the wrong styling. The rounded corners should be on the opposite side

Screen Shot 2019-08-08 at 7 36 33 PM

**Comment by InnaAtanasova** [comment]
The documentation is not showing validation for Datetime Picker

Screen Shot 2019-08-08 at 7 53 46 PM

**Comment by InnaAtanasova** [comment]
Prob it's intended and we need to discuss it in terms of UX, but I noticed that when you change the date/time in the input field you need to press `Enter` to update the object. If you just type the desired date for example, and click outside, the popover body closes, the input field "remembers" the user's input but the datetime object is not updated. We should either clear the input (or reset it to what it was), or update the object's value, or something else...

Screen Shot 2019-08-08 at 7 59 51 PM

**Comment by JKMarkowski** [comment]
Hi @InnaAtanasova, All your comments are solved :)

**Pull request merged**